### PR TITLE
feat/brightcove-finish-evt

### PR DESCRIPTION
**Pull request by dickie81**
Finish property added to change events to indicate when video has finished playing. This is for lifecycle management in native implementations.

Also makes playheadPosition report more consistently across platforms.

### CODE REVIEW

File: packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m [hunk 1]
```diff
@@ -176,6 +187,10 @@ - (void)playbackController:(id<BCOVPlaybackController>)controller playbackSessio
 
   if ([kBCOVPlaybackSessionLifecycleEventPause isEqualToString:lifecycleEvent.eventType]) {
     _playerStatus = @"paused";
+    
+    if([_playheadPosition isEqualToNumber:_duration]) {
```
Comment (mabidakun) [inline packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: **Beware**: Media players are famous for not quite achieving this kind of equality, especially given that `duration` is a floating point value.

Comment (dickie81) [reply packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: Tested with a few videos without issue - have you got a better suggestion?

File: packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m [hunk 2]
```diff
@@ -185,10 +200,21 @@ - (void)playbackController:(id<BCOVPlaybackController>)controller playbackSessio
...
   NSNumber *progressNumber = [NSNumber numberWithDouble:progress];
 
-  _playheadPosition = [progressNumber stringValue];
+  NSComparisonResult compareZeroResult = [_zero compare:progressNumber];
+  NSComparisonResult compareDurationResult = [_duration compare:progressNumber];
+  
+  if (compareZeroResult == NSOrderedDescending) { // result less than zero - default to zero
+    _playheadPosition = _zero;
+  } else if (compareDurationResult == NSOrderedAscending) { // result greater than duration - default to duration
+    _playheadPosition = _duration;
```
Comment (dickie81) [inline packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: @mabidakun if the video has finished `progressNumber` is positive infinity so `_playheadPosition` is set to `_duration` which should make `[_playheadPosition isEqualToNumber:_duration]` always work?

**Comment by trevorah** [comment]
Why cant you use integers? Android duration seems to be in millis, and ios duration is in seconds.

File: packages/brightcove-video/android/src/main/java/uk/co/news/rntbrightcovevideo/BrightcovePlayerView.java [hunk 3]
```diff
@@ -31,32 +31,52 @@ public BrightcovePlayerView(final Context context) {
...
         });
-
-
         eventEmitter.on(EventType.PLAY, onEvent("playing"));
         eventEmitter.on(EventType.PAUSE, onEvent("paused"));
+        eventEmitter.on(EventType.PROGRESS, onEvent("playing"));
+        eventEmitter.on(EventType.COMPLETED, new EventListener() {
+            @Override
+            public void processEvent(Event e) {
+                playerView.bubbleState("paused", playerView.getDurationSecs());
```
Comment (chrishoyes) [inline packages/brightcove-video/android/src/main/java/uk/co/news/rntbrightcovevideo/BrightcovePlayerView.java]: is this correct?

**Comment by dickie81** [comment]
Playhead position now returned in whole milliseconds across all platforms.

File: packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m [hunk 4]
```diff
@@ -142,7 +147,12 @@ - (void)emitStatus {
     return;
   }
 
-  self.onChange(@{@"playerStatus": _playerStatus, @"playheadPosition": _playheadPosition});
+  self.onChange(@{
+    @"playerStatus": _playerStatus,
+    @"playheadPosition": [NSNumber numberWithLong:_playheadPosition],
+    @"duration": [NSNumber numberWithLong:_duration],
+    @"finished": [NSNumber numberWithBool:_finished]
```
Comment (trevorah) [inline packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: why does finished have to be a number?

File: packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m [hunk 5]
```diff
@@ -16,8 +16,10 @@ @interface RNTBrightcove () <BCOVPlaybackControllerDelegate, BCOVPUIPlayerViewDe
 @implementation RNTBrightcove {
   RCTEventDispatcher *_eventDispatcher;
   NSString *_playerStatus;
-  NSString *_playheadPosition;
+  NSInteger _playheadPosition;
```
Comment (trevorah) [inline packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: `_progressInMillis` would make this code a lot more readable.

File: packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m [hunk 6]
```diff
@@ -177,6 +187,10 @@ - (void)playbackController:(id<BCOVPlaybackController>)controller playbackSessio
   if ([kBCOVPlaybackSessionLifecycleEventPause isEqualToString:lifecycleEvent.eventType]) {
     _playerStatus = @"paused";
 
+    if(_playheadPosition == _duration) {
+      _finished = YES;
+    }
```
Comment (trevorah) [inline packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: `_playheadPosition >= _duration`, just to be sure.

File: packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m [hunk 7]
```diff
@@ -186,9 +200,15 @@ - (void)playbackController:(id<BCOVPlaybackController>)controller playbackSessio
...
+  _finished = NO;
 
-  _playheadPosition = [progressNumber stringValue];
+  if (progress == -INFINITY) {
+    _playheadPosition = 0;
+  } else if (progress == INFINITY) {
+    _playheadPosition = _duration;
+  } else {
+    _playheadPosition = progress * 1000;
+  }
```
Comment (trevorah) [inline packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: Can you add some comments about why there are infinities floating about please?

File: packages/brightcove-video/brightcove-video.stories.js [hunk 8]
```diff
@@ -293,6 +293,14 @@ storiesOf("BrightcoveVideo", module)
       onChange={firstArgJSONAction("change")}
     />
   ))
+  .add("BrightcoveVideo with event listener", () => (
+    <BrightcoveVideo
+      policyKey={policyKey}
+      videoId={videoId}
+      accountId={accountId}
+      onChange={firstArgJSONAction("change")}
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.stories.js]: why is `onChange` exposed? Shouldnt that be `nativeOnly` and the js side exposes the nicer events?

File: packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m (hunk 4 above)
Comment (dickie81) [reply packages/brightcove-video/ios/RNTBrightcove/RNTBrightcove.m]: It gives a warning if I don't assign an object here. I'm not experienced enough with obj-c to know why.

**Pull request merged**